runtime/expr/agg/map.go

```diff
@@ -0,0 +1,98 @@
+package agg
+
+import (
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/zcode"
+)
+
+type Map struct {
+	entries map[string]mapEntry
+	values  []zed.Value
+	size    int
+	scratch []byte
+}
+
+func newMap() *Map {
+	return &Map{entries: make(map[string]mapEntry)}
+}
+
+var _ Function = (*Collect)(nil)
+
+type mapEntry struct {
+	key *zed.Value
+	val *zed.Value
+}
+
+func (m *Map) Consume(val *zed.Value) {
+	if val.IsNull() {
+		return
+	}
+	mtyp, ok := zed.TypeUnder(val.Type).(*zed.TypeMap)
+	if !ok {
+		return
+	}
+	it := val.Bytes.Iter()
+	for !it.Done() {
+		ktyp, kb := untagUnionValue(mtyp.KeyType, it.Next())
+		vtyp, vb := untagUnionValue(mtyp.ValType, it.Next())
+		m.scratch = zed.AppendTypeValue(m.scratch[:0], ktyp)
+		m.scratch = append(m.scratch, kb...)
```

Comment:
Using kb here means this can't distinguish between the null key and the zero-length key. Here's one way to do that:

```go
	keyTagAndBody := it.NextTagAndBody()
	ktyp, kb := untagUnionValue(mtyp.KeyType, keyTagAndBody.Body())
	m.scratch = zed.AppendTypeValue(m.scratch[:0], ktyp)
	m.scratch = append(m.scratch, keyTagAndBody...)
```

```